sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -613,13 +633,62 @@ def _blocking_request(self, request):
     if response.error:
       raise RuntimeError(response.error)
     else:
+      # TODO mxm here we could update the cache token
       return response
 
   def _next_id(self):
     self._last_id += 1
     return str(self._last_id)
 
 
+class CachingGrpcStateHandler(GrpcStateHandler):
+  """ TODO mxm """
+
+  def __init__(self, state_stub, global_state_cache):
+    super(CachingGrpcStateHandler, self).__init__(state_stub)
+    self._state_cache = global_state_cache
+
+  def blocking_get(self, state_key, coder):
+    cache_tokens = self._context.cache_tokens
+    if not cache_tokens:
+      # no cache tokens, can't do a lookup in the cache
+      return super(CachingGrpcStateHandler, self).blocking_get(
+          state_key, coder)
+    cache_state_key = CachingGrpcStateHandler.convert_to_cache_key(state_key)
+    value = self._state_cache.get(cache_state_key, cache_tokens)
+    if not value:
+      # Cache miss, need to retrieve from the Runner
+      value = super(CachingGrpcStateHandler, self).blocking_get(
+          state_key, coder)
+      # TODO mxm uses always the first cache token for now
+      self._state_cache.put(
+          CachingGrpcStateHandler.convert_to_cache_key(state_key),
+          cache_tokens[0], value)
+    return value
+
+  def blocking_append(self, state_key, data):
+    cache_tokens = self._context.cache_tokens
+    self._state_cache.put(
```

Comment:
if we are considering the current protocol then put method should return the new token, in that case, we should caching it after updating the runner.

Reply:
I've already implemented the last design discussion around async handling where we assume validity of the Runner-provided cache tokens for the duration of a bundle. The async handling itself is still a TODO.

Reply:
With the new version, we don't return the token anymore on append/clear.